fraud/service.go

```diff
@@ -0,0 +1,132 @@
+package fraud
+
+import (
+	"context"
+	"encoding/hex"
+	"errors"
+	"sync"
+
+	pubsub "github.com/libp2p/go-libp2p-pubsub"
+)
+
+// service is propagating and validating Fraud Proofs
+// service implements Service interface.
+type service struct {
+	topics *topics
+
+	unmarshalers map[ProofType]ProofUnmarshaler
+	mu           sync.RWMutex
+
+	getter headerFetcher
+}
+
+func NewService(p *pubsub.PubSub, getter headerFetcher) Service {
+	return &service{
+		topics: &topics{
+			pubsub:       p,
+			pubSubTopics: make(map[ProofType]*pubsub.Topic),
+		},
+		unmarshalers: make(map[ProofType]ProofUnmarshaler),
+		getter:       getter,
+	}
+}
+
+func (f *service) Subscribe(proofType ProofType) (Subscription, error) {
+	// TODO: @vgonkivs check if fraud proof is in fraud store, then return with error
+	f.mu.RLock()
+	u, ok := f.unmarshalers[proofType]
+	f.mu.RUnlock()
+	if !ok {
+		return nil, errors.New("fraud: unmarshaler is not registered")
+	}
+
+	t, wasNotjoined, err := f.topics.getTopic(proofType)
+	if err != nil {
+		return nil, err
+	}
+	// if topic was joined for the first time then we should register a validator for it.
+	if wasNotjoined {
+		if err = f.topics.registerValidator(proofType, f.processIncoming); err != nil {
+			return nil, err
+		}
+	}
+
+	return newSubscription(t, u)
+}
+
+func (f *service) RegisterUnmarshaler(proofType ProofType, u ProofUnmarshaler) error {
+	f.mu.Lock()
+	defer f.mu.Unlock()
+
+	if _, ok := f.unmarshalers[proofType]; ok {
+		return errors.New("fraud: unmarshaler is already registered")
+	}
+	f.unmarshalers[proofType] = u
+
+	return nil
+}
+
+func (f *service) UnregisterUnmarshaler(proofType ProofType) error {
```

Comment:
Again I do not know where this will be used and why it's even necessary. IMO, the things that go together should be grouped together and not allowed to be "registered" unless the entire package is there. That means

A topic needs a validator and needs an unmarshaler. Those things should be all required at the same time in order for the blob (topic+unmarshaler+validator) to be registered on the fraud service.

I think this part can be done better. Maybe lets have a call on it tomorrow.